### PR TITLE
AdSecGh-64 spike testing component cherry

### PR DESCRIPTION
**Pull request by psarras**
This PR is attempting to show how you can test a single component, namely `FlattenRebar.cs`

### CODE REVIEW

File: AdSecGH/Components/3_Section/EditSection.cs [hunk 1]
```diff
@@ -52,26 +55,32 @@ protected override void RegisterOutputParams(GH_OutputParamManager pManager) {
...
       if (in_section == null) {
-        AddRuntimeMessage(GH_RuntimeMessageLevel.Warning, "Input parameter " + Params.Input[0].NickName + " failed to collect data!");
+        AddRuntimeMessage(GH_RuntimeMessageLevel.Warning,
+          "Input parameter " + Params.Input[0].NickName + " failed to collect data!");
         return;
       }
+
       // 1 profile
       AdSecProfileGoo profile = null;
       if (Params.Input[1].SourceCount > 0) {
```
Comment (DominikaLos) [inline AdSecGH/Components/3_Section/EditSection.cs:78]: Params.Input[1] if null then you will got exception

File: AdSecGH/Components/3_Section/EditSection.cs [hunk 2]
```diff
@@ -52,26 +55,32 @@ protected override void RegisterOutputParams(GH_OutputParamManager pManager) {
...
         profile = AdSecInput.AdSecProfileGoo(this, DA, 1, true);
       } else {
         profile = new AdSecProfileGoo(in_section.Section.Profile, in_section.LocalPlane);
       }
+
       DA.SetData(1, profile);
```
Comment (DominikaLos) [inline AdSecGH/Components/3_Section/EditSection.cs:84]: why not to set all data at the end of the method?

File: AdSecGH/Components/3_Section/EditSection.cs [hunk 3]
```diff
@@ -89,6 +98,7 @@ protected override void SolveInstance(IGH_DataAccess DA) {
       } else {
```
Comment (DominikaLos) [inline AdSecGH/Components/3_Section/EditSection.cs:98]: as above

File: AdSecGH/Components/3_Section/EditSection.cs [hunk 4]
```diff
@@ -99,17 +109,19 @@ protected override void SolveInstance(IGH_DataAccess DA) {
       if (Params.Input[4].SourceCount > 0) {
         reinforcements = AdSecInput.ReinforcementGroups(this, DA, 4, true);
       } else {
-        foreach (IGroup rebarGrp in in_section.Section.ReinforcementGroups) {
+        foreach (var rebarGrp in in_section.Section.ReinforcementGroups) {
```
Comment (DominikaLos) [inline AdSecGH/Components/3_Section/EditSection.cs:112]: Section?. use null check pls

File: AdSecGH/Components/3_Section/FlattenRebar.cs [hunk 5]
```diff
@@ -1,130 +1,139 @@
...
-
-    public FlattenRebar() : base(
-      "FlattenRebars",
-      "FRb",
-      "Flatten all rebars in a section into single bars.",
-      CategoryName.Name(),
-      SubCategoryName.Cat4()) {
-      Hidden = true;
+
+  public class FlattenRebarGhComponent : FlattenRebarComponent {
```
Comment (DominikaLos) [inline AdSecGH/Components/3_Section/FlattenRebar.cs:20]: here you have FlattenRebarGhComponent, and for previous file is just EditSection. Why the difference?

File: AdSecGH/Components/3_Section/FlattenRebar.cs [hunk 6]
```diff
@@ -1,130 +1,139 @@
...
+      Hidden = true;
     }
+
+    public override Guid ComponentGuid => new Guid("879ecac6-464d-4286-9113-c15fcf03e4e6");
+
+    public override GH_Exposure Exposure => GH_Exposure.tertiary | GH_Exposure.obscure;
+    public override OasysPluginInfo PluginInfo => AdSecGH.PluginInfo.Instance;
+    protected override Bitmap Icon => Resources.FlattenRebar;
+
+    // protected override void SolveInstance(IGH_DataAccess DA) {
```
Comment (DominikaLos) [inline AdSecGH/Components/3_Section/FlattenRebar.cs:81]: remove comments :)

File: AdSecGH/Components/6_Results/RebarStressStrain.cs [hunk 7]
```diff
@@ -70,17 +78,21 @@ protected override void SolveInstance(IGH_DataAccess DA) {
...
-        AddRuntimeMessage(GH_RuntimeMessageLevel.Warning, "Input parameter " + Params.Input[1].NickName + " failed to collect data!");
+        AddRuntimeMessage(GH_RuntimeMessageLevel.Warning,
+          "Input parameter " + Params.Input[1].NickName + " failed to collect data!");
         return;
       }
 
       // create flattened section to extract rebars
       ISection flat = null;
-      if (solution.m_section.DesignCode != null) //{ code = Oasys.AdSec.DesignCode.EN1992.Part1_1.Edition_2004.NationalAnnex.NoNationalAnnex; }
+      if (
```
Comment (DominikaLos) [inline AdSecGH/Components/6_Results/RebarStressStrain.cs:93]: very strange formatting XD

File: AdSecGH/Helpers/AdSecInput.cs [hunk 8]
```diff
@@ -357,13 +429,19 @@ internal static List<AdSecRebarGroup> ReinforcementGroups(GH_Component owner, IG
...
-        owner.AddRuntimeMessage(GH_RuntimeMessageLevel.Warning, "Input parameter " + owner.Params.Input[inputid].NickName + " failed to collect data!");
       }
+
+      if (!isOptional) {
+        owner.AddRuntimeMessage(GH_RuntimeMessageLevel.Warning,
+          "Input parameter " + owner.Params.Input[inputid].NickName + " failed to collect data!");
+      }
+
       return null;
     }
```
Comment (DominikaLos) [inline AdSecGH/Helpers/AdSecInput.cs:447]: remove comments

File: AdSecGH/Helpers/AdSecInput.cs [hunk 9]
```diff
@@ -465,17 +566,25 @@ internal static Oasys.Collections.IList<IStressStrainPoint> StressStrainPoints(G
...
```
Comment (DominikaLos) [inline AdSecGH/Helpers/AdSecInput.cs:590]: as above

File: AdSecGH/Helpers/AdSecInput.cs [hunk 10]
```diff
@@ -532,15 +649,20 @@ internal static Oasys.Profiles.IWeb Web(GH_Component owner, IGH_DataAccess DA, i
...
           return null;
         }
       } else if (!isOptional) {
-        owner.AddRuntimeMessage(GH_RuntimeMessageLevel.Warning, "Input parameter " + owner.Params.Input[inputid].NickName + " failed to collect data!");
+        owner.AddRuntimeMessage(GH_RuntimeMessageLevel.Warning,
+          "Input parameter " + owner.Params.Input[inputid].NickName + " failed to collect data!");
       }
+
       return null;
     }
```
Comment (DominikaLos) [inline AdSecGH/Helpers/AdSecInput.cs:668]: as above

File: AdSecGH/Helpers/BusinessExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+        }, {
+          typeof(StringArrayParam), a => new Param_String {
+            Name = a.Name,
+            NickName = a.NickName,
+            Description = a.Description,
+            Access = GetAccess(a),
+          }
+        },
+      };
+    // private static readonly Dictionary<Type, Func<object, Attribute>> FromGoo
```
Comment (DominikaLos) [inline AdSecGH/Helpers/BusinessExtensions.cs:81]: as above

File: AdSecGH/Helpers/BusinessExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+namespace Oasys.GH.Helpers {
+
+  public class IAdSecSectionParameter : ParameterAttribute<AdSecSectionGoo> { }
+
+  public class AdSecPointArrayParameter : BaseArrayParameter<AdSecPointGoo> { }
+  public class AdSecMaterialArrayParam : BaseArrayParameter<AdSecMaterialGoo> { }
+
+  public static class BusinessExtensions {
+
+    private static readonly Dictionary<Type, Func<Attribute, IGH_Param>> ToGhParam
```
Comment (DominikaLos) [inline AdSecGH/Helpers/BusinessExtensions.cs:28]: interesting idea! nice!

File: AdSecGHTests/Components/FlattenRebarGhTests.cs [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+  public class FlattenRebarGhTests {
+    private readonly FlattenRebarGhComponent component;
+    private readonly IDesignCode designCode = IS456.Edition_2000;
+    private bool valueChanged;
+
+    public FlattenRebarGhTests() {
+      component = new FlattenRebarGhComponent();
+    }
+
+    private ISection Section => new SectionBuilder().WithWidth(40).CreateSquareSection().Build();
```
Comment (DominikaLos) [inline AdSecGHTests/Components/FlattenRebarGhTests.cs:32]: properties before constructor pls

File: AdSecGH/Components/3_Section/EditSection.cs (hunk 4 above)
Comment (psarras) [reply AdSecGH/Components/3_Section/EditSection.cs:112]: these file are from autoformat, out of scope for this PR

File: AdSecGH/Components/3_Section/EditSection.cs (hunk 3 above)
Comment (psarras) [reply AdSecGH/Components/3_Section/EditSection.cs:98]: these file are from autoformat, out of scope for this PR

File: AdSecGH/Components/3_Section/EditSection.cs (hunk 1 above)
Comment (psarras) [reply AdSecGH/Components/3_Section/EditSection.cs:78]: this file are from autoformat, out of scope for this PR

File: AdSecGH/Components/3_Section/EditSection.cs (hunk 2 above)
Comment (psarras) [reply AdSecGH/Components/3_Section/EditSection.cs:84]: this file are from autoformat, out of scope for this PR

File: AdSecGH/Helpers/AdSecInput.cs (hunk 8 above)
Comment (psarras) [reply AdSecGH/Helpers/AdSecInput.cs:447]: which comment?

File: AdSecGH/Helpers/AdSecInput.cs (hunk 9 above)
Comment (psarras) [reply AdSecGH/Helpers/AdSecInput.cs:590]: ?

File: AdSecGH/Helpers/AdSecInput.cs (hunk 10 above)
Comment (psarras) [reply AdSecGH/Helpers/AdSecInput.cs:668]: ?

File: AdSecGH/Components/3_Section/FlattenRebar.cs (hunk 5 above)
Comment (psarras) [reply AdSecGH/Components/3_Section/FlattenRebar.cs:21]: could you elaborate?

File: AdSecGH/Helpers/AdSecInput.cs (hunk 8 above)
Comment (DominikaLos) [reply AdSecGH/Helpers/AdSecInput.cs:447]: from line 448

File: AdSecGH/Helpers/AdSecInput.cs (hunk 9 above)
Comment (DominikaLos) [reply AdSecGH/Helpers/AdSecInput.cs:590]: from line 591

File: AdSecGH/Helpers/AdSecInput.cs (hunk 10 above)
Comment (DominikaLos) [reply AdSecGH/Helpers/AdSecInput.cs:668]: from line 669

File: AdSecGH/Components/3_Section/FlattenRebar.cs (hunk 6 above)
Comment (psarras) [reply AdSecGH/Components/3_Section/FlattenRebar.cs:82]: fixed

File: AdSecGH/Components/3_Section/FlattenRebar.cs (hunk 5 above)
Comment (psarras) [reply AdSecGH/Components/3_Section/FlattenRebar.cs:21]: will think of something better

File: AdSecGHTests/Components/FlattenRebarGhTests.cs (hunk 13 above)
Comment (psarras) [reply AdSecGHTests/Components/FlattenRebarGhTests.cs:32]: I had a different rule on my settings. I made it a field though

File: AdSecCoreTests/FlattenRebarTests.cs [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    [Fact]
+    public void ShouldAddOutputsAsManyDiametersAsRebars() {
+      Assert.Single(operation.Diameter.Value);
+    }
+
+    [Fact]
+    public void ShouldAddOutputsAsManyPositionsAsRebars() {
+      Assert.Single(operation.Position.Value);
```
Comment (DominikaLos) [inline AdSecCoreTests/FlattenRebarTests.cs:49]: shouldn't you compare count values?

File: AdSecGH/Components/3_Section/FlattenRebar.cs [hunk 15]
```diff
@@ -1,130 +1,82 @@
...
+      };
+      var adSecSection = AdSecSection as Attribute;
+      FromAttribute(ref adSecSection, Section);
+      var adSecPoint = AdSecPoint as Attribute;
+      FromAttribute(ref adSecPoint, Position);
+      AdSecPoint.Name = Position.NameWithUnits(DefaultUnits.LengthUnitGeometry);
+      Diameter.Name = Diameter.NameWithUnits(DefaultUnits.LengthUnitGeometry);
     }
 
+    public AdSecPointArrayParameter AdSecPoint { get; set; } = new AdSecPointArrayParameter();
```
Comment (DominikaLos) [inline AdSecGH/Components/3_Section/FlattenRebar.cs:40]: not sure it should be here, maybe in an attribute class?

File: AdSecGH/Components/6_Results/RebarStressStrain.cs [hunk 16]
```diff
@@ -21,43 +24,48 @@
...
+      pManager.AddGenericParameter("SLS Strain [" + strainUnitAbbreviation + "]", "εk",
+        "SLS strain for each rebar position", GH_ParamAccess.list);
+      pManager.AddGenericParameter("SLS Stress [" + stressUnitAbbreviation + "]", "σk",
+        "SLS stress for each rebar position", GH_ParamAccess.list);
     }
 
     protected override void SolveInstance(IGH_DataAccess DA) {
       // get solution input
-      AdSecSolutionGoo solution = AdSecInput.Solution(this, DA, 0);
+      var solution = AdSecInput.Solution(this, DA, 0);
```
Comment (DominikaLos) [inline AdSecGH/Components/6_Results/RebarStressStrain.cs:65]: do not use var if type of variable is not clear

File: AdSecGH/Helpers/BusinessGlue.cs [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+using OasysGH;
+using OasysGH.Components;
+
+namespace Oasys.GH.Helpers {
+
+  public abstract class ComponentAdapter<T> : GH_OasysComponent, IDefaultValues where T : IFunction {
+
+    public readonly T BusinessComponent = Activator.CreateInstance<T>();
+
+    public ComponentAdapter() : base("", "", "", "", "") {
```
Comment (DominikaLos) [inline AdSecGH/Helpers/BusinessGlue.cs:17]: string.empty

File: AdSecGH/Helpers/BusinessGlue.cs [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+      NickName = BusinessComponent.Metadata.NickName;
+      Description = BusinessComponent.Metadata.Description;
+      Category = BusinessComponent.Organisation.Category;
+      SubCategory = BusinessComponent.Organisation.SubCategory;
+    }
+
+    public override OasysPluginInfo PluginInfo { get; } = AdSecGH.PluginInfo.Instance;
+    public void SetDefaultValues() { BusinessComponent.SetDefaultValues(this); }
+
+    protected override void RegisterInputParams(GH_InputParamManager pManager) {
```
Comment (DominikaLos) [inline AdSecGH/Helpers/BusinessGlue.cs:28]: is pManager needed here?

File: AdSecGH/Helpers/BusinessGlue.cs [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    protected override void SolveInstance(IGH_DataAccess DA) {
+      BusinessComponent.UpdateInputValues(this, DA);
+      BusinessComponent.Compute();
+      BusinessComponent.SetOutputValues(this, DA);
+    }
+  }
+
+  public interface IDefaultValues {
```
Comment (DominikaLos) [inline AdSecGH/Helpers/BusinessGlue.cs:43]: let's move to another file

File: AdSecGH/Helpers/BusinessGlue.cs [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    public override OasysPluginInfo PluginInfo { get; } = AdSecGH.PluginInfo.Instance;
+    public void SetDefaultValues() { BusinessComponent.SetDefaultValues(this); }
+
+    protected override void RegisterInputParams(GH_InputParamManager pManager) {
+      BusinessComponent.PopulateInputParams(this);
+    }
+
+    protected override void RegisterOutputParams(GH_OutputParamManager pManager) {
```
Comment (DominikaLos) [inline AdSecGH/Helpers/BusinessGlue.cs:32]: is pManager needed here?

File: AdSecGH/Helpers/BusinessGlue.cs [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+      BusinessComponent.Compute();
+      BusinessComponent.SetOutputValues(this, DA);
+    }
+  }
+
+  public interface IDefaultValues {
+    void SetDefaultValues();
+  }
+
+  public abstract class DropdownAdapter<T> : GH_OasysDropDownComponent, IDefaultValues where T : IFunction {
```
Comment (SandeepArup) [inline AdSecGH/Helpers/BusinessGlue.cs:47]: @psarras I do not have any strong reservations if something is working, but people generally follow one project file (.cs) per class.

File: AdSecGH/Components/3_Section/FlattenRebar.cs [hunk 22]
```diff
@@ -1,130 +1,82 @@
...
-      "FRb",
-      "Flatten all rebars in a section into single bars.",
-      CategoryName.Name(),
-      SubCategoryName.Cat4()) {
-      Hidden = true;
+
+  public class FlattenRebarGh : FlattenRebarFunction {
+
+    public FlattenRebarGh() {
+      AdSecSection.OnValueChanged += goo => {
```
Comment (SandeepArup) [inline AdSecGH/Components/3_Section/FlattenRebar.cs:26]: I have not run the code, but how is this event useful here? Can we not simply use
if (AdSecSection.Value != null) {
  Section.Value = AdSecSection.Value.Value.Section;
}

File: AdSecGH/Helpers/BusinessGlue.cs (hunk 21 above)
Comment (DominikaLos) [reply AdSecGH/Helpers/BusinessGlue.cs:47]: agree

File: AdSecGH/Components/3_Section/FlattenRebar.cs (hunk 15 above)
Comment (psarras) [reply AdSecGH/Components/3_Section/FlattenRebar.cs:40]: I understand now what you meant. I moved the other method to the Attribute class

**Review by DominikaLos — APPROVED** [review]

**Review by SandeepArup — APPROVED** [review]

**Pull request merged**